python/across_api/auth/api.py

```diff
@@ -78,15 +83,41 @@ async def __call__(self, request: Request) -> None:
         else:
             raise AssertionError("No credentials passed.")
 
+        return {}
+
 
-security = JWTBearer(
+JWTBearerSecurity = JWTBearer(
     scheme_name="ACROSS API Authorization",
     description="Enter your access token.",
 )
-JWTBearerDep = [Depends(security)]
 
 
-@app.get("/auth/verify", dependencies=JWTBearerDep)
+async def ScopeAuthenticate(
+    security_scopes: SecurityScopes,
+    access_token: Annotated[dict, Depends(JWTBearerSecurity)],
+):
+    valid = False
+    scopes = access_token.get("scope", "")
+
+    # assuming the jwt scopes will be comma separated
+    token_scopes = scopes.split(",")
+
+    # check if user scopes is in endpoint scope
+    if len(security_scopes.scopes):
+        valid = any(scope in security_scopes.scopes for scope in token_scopes)
+    else:  # endpoint has no scopes
+        valid = True
```

Comment:
```suggestion
    valid = all(scope in token_scopes for scope in security_scopes.scopes)
```

Reply:
Oops, sorry for repeating myself. I am looking at your comment https://github.com/nasa-gcn/gcn.nasa.gov/pull/1874#discussion_r1478502298 again.

Reply:
Oh, right. I think that this version addresses your comment.

Reply:
This will work if our endpoints only have one scope. The `any()` will work if the endpoint's `security_scopes.scopes` contains more than one.

I'm fine with changing it, but if we envision our endpoints having more than one scope, the `all()` will return false.

Reply:
Huh? This is supposed to check that all of the required scopes are present in the access token. Why would that fail if the endpoints require more than one scope?

Reply:
Our endpoints can be decorated with a list of scopes. For example `/read_foo` api endpoint can be decorated with `scopes = ["read:/", "read:/foo"]`. There might be an instance where a user only has a JWT with scope privileges that only contain ["read:/foo"].

This leads to:
```python
security_scopes.scopes =  ["read:/", "read:/foo"] #api endpoint scopes
token_scopes = ["read:/foo"]

valid = all(scope in token_scopes for scope in security_scopes.scopes) #returns False
``` 

`valid` returns `False` with this implementation, event though the user scopes are within the endpoint scopes.

whereas if we implement this validation with ` any(scope in token_scopes for scope in security_scopes.scopes)` it will return `valid = True`, which is the desired outcome. The only issue is that if `security_scopes.scopes` is empty, then we will need to check that since the `any(...)` evaluation will return `False`.

Reply:
I see. However, the `all(...)` expression I gave is consistent with the behavior of the FastAPI example code in https://fastapi.tiangolo.com/advanced/security/oauth2-scopes/#global-view.

Reply:
Is there some documentation about SciTokens that says that `read:/` is supposed to imply `read:/foo` as well? If this is supposed to model POSIX permissions, they don't work that way.

Also, if we intend a token to have both `read:/` and `read:/foo` permissions, couldn't we just issue a token with both scopes?

Reply:
I see what you mean in their tutorial. 

Here is the documentation on [SciToken Claims](https://scitokens.org/technical_docs/Claims). Go down to the section on "SciToken Claim Semantics". 

I assume the `read:/` is an administrator role to _read everything_ where `read:/foo" is a restricted read, where the user only has the ability to _read what is in foo_. I 

Your vision is that when a user is assigned a scope that includes `read:/foo` that it will automatically give them the scopes of `["read:/", "read:/foo"]`? Which would work with the `all(...)` implementation

Reply:
> Your vision is that when a user is assigned a scope that includes `read:/foo` that it will automatically give them the scopes of `["read:/", "read:/foo"]`? Which would work with the `all(...)` implementation

No, following the example in the "SciTokens Scopes" section, it should be the opposite: an assigned scope of `read:/` should be equivalent to `["read:/", "read:/foo"]`.

Anyway, I don't want to get into path parsing in this PR, though we could in a future one.

Reply:
Ok. I've implemented the `all(...)`, I still had to check if the endpoint scopes had a length because the line of code would return `False` if the `security_scopes.scopes` list was empty.